### PR TITLE
[common] add `BinarySearch` module and `Stringfy`

### PR DESCRIPTION
**Pull request by abtink**
This commit adds set of generic binary search related functions in
OpenThread core under `BinarySearch` module. The binary search module
replaces and enhances the existing `Lookup` table module (which was
limited to search based on strings and mainly used by CLI to map from
command string to a handler function).

This commit also adds a new class `Stringfy` in `string.hpp` to help
convert from a set of `uint16_t` values (e.g., a non-sequential
`enum`) to strings using binary search in a table. The new methods are
used for converting `Coap::Code` and `IpProto` enumerations to
corresponding strings.

### CODE REVIEW

File: src/core/common/binary_search.cpp [hunk 1]
```diff
@@ -56,17 +53,17 @@ const void *LookupTable::Find(const char *aName,
...
+        // only its size is given as `aEntrySize`. Based on this, we
+        // can calculate the pointer to the table entry at any index
         // (such as `[middle]`) which is then passed to the given
-        // `aNameGetter` function which knows how to cast the void
-        // pointer to proper entry type and get the enclosed `mName`
-        // field. This model keeps the implementation generic and
-        // re-usable allowing it to be used with any entry type.
+        // `aComparator` function which knows how to cast the void
+        // pointer to proper entry type and compare it with `aKey`.
+        // This model keeps the implementation generic and re-usable
```
Comment (jwhui) [inline src/core/common/binary_search.cpp:61]: nit:
```suggestion
        // This model keeps the implementation generic and reusable
```

File: src/core/common/binary_search.hpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+     * @tparam Entry       The table `Entry` type.
+     * @tparam kLength     The array length (number of entries in the array).
+     *
+     * The `Entry` class MUST provide the following method to compare the entry against a given key.
+     *
+     *    int Entry::Compare(const Key &aKey) const;
+     *
+     * The return value indicates the comparison result between `aKey` and the entry (similar to `strcmp()`), i.e.,
+     * zero means perfect match, positive (> 0) indicates `aKey` is larger than entry, and negative indicates `aKey` is
+     * smaller than entry.
```
Comment (jwhui) [inline src/core/common/binary_search.hpp:61]:
```suggestion
     * The return value indicates the comparison result between @p aKey and the entry (similar to `strcmp()`), i.e.,
     * zero means perfect match, positive (> 0) indicates @p aKey is larger than entry, and negative indicates @p aKey is
     * smaller than entry.
```

File: src/core/common/string.hpp [hunk 3]
```diff
@@ -204,6 +205,51 @@ char ToUppercase(char aChar);
...
+ * @retval TRUE   The sequence is a valid UTF-8 string.
+ * @retval FALSE  The sequence is not a valid UTF-8 string.
+ *
+ */
+bool IsValidUtf8String(const char *aString, size_t aLength);
+
+/**
+ * This `constexpr` function checks whether two given C strings are in order (alphabetical order).
+ *
+ * This is intended for use from `static_asstert`, e.g., checking if a lookup table entries are sorted. It is not
```
Comment (jwhui) [inline src/core/common/string.hpp:236]:
```suggestion
 * This is intended for use from `static_assert`, e.g., checking if a lookup table entries are sorted. It is not
```

File: src/core/common/binary_search.hpp (hunk 2 above)
Comment (abtink) [reply src/core/common/binary_search.hpp]: I changed this.

One thing related to `@p`: I somehow always assumed that it indicates (in doxygen) that the next term is actually a parameter. I think we use it in OT more or less under this assumption.
 
I learned recently that `@p` may just be a formatting indicator and practically same as using backtick in doxygen.

**Review by jwhui — APPROVED** [review]

**Pull request merged**